crates/sui-indexer-alt-e2e-tests/src/lib.rs

```diff
@@ -1,6 +1,387 @@
+        client_args: ClientArgs,
+        system_package_task_args: SystemPackageTaskArgs,
+        indexer_config: IndexerConfig,
+        rpc_config: RpcConfig,
+        registry: &prometheus::Registry,
+        cancel: CancellationToken,
+    ) -> anyhow::Result<Self> {
+        let rpc_port = get_available_port();
+        let rpc_listen_address = SocketAddr::new(IpAddr::V4(Ipv4Addr::LOCALHOST), rpc_port);
+
+        let database = TempDb::new().context("Failed to create database")?;
+
+        let db_args = DbArgs {
+            database_url: database.database().url().clone(),
+            ..Default::default()
+        };
+
+        let rpc_args = RpcArgs {
+            rpc_listen_address,
+            ..Default::default()
+        };
+
+        let db = Db::for_read(db_args.clone())
+            .await
+            .context("Failed to connect to database")?;
+
+        let with_genesis = true;
+        let indexer = setup_indexer(
+            db_args.clone(),
+            indexer_args,
+            client_args,
+            indexer_config,
+            with_genesis,
+            registry,
+            cancel.child_token(),
+        )
+        .await
+        .context("Failed to setup indexer")?;
+
+        let pipelines = indexer.pipelines().collect();
+        let indexer = indexer.run().await.context("Failed to start indexer")?;
+
+        let jsonrpc = start_rpc(
+            db_args,
+            rpc_args,
+            system_package_task_args,
+            rpc_config,
+            registry,
+            cancel.child_token(),
+        )
+        .await
+        .context("Failed to start JSON-RPC server")?;
+
+        Ok(Self {
+            rpc_listen_address,
+            db,
+            pipelines,
+            indexer,
+            jsonrpc,
+            database,
+            cancel,
+        })
+    }
+
+    /// The URL to talk to the database on.
+    pub fn db_url(&self) -> Url {
+        self.database.database().url().clone()
+    }
+
+    /// The URL to send JSON-RPC requests to.
+    pub fn rpc_url(&self) -> Url {
+        Url::parse(&format!("http://{}/", self.rpc_listen_address))
+            .expect("Failed to parse RPC URL")
+    }
+
+    /// Returns the latest checkpoint that we have all data for in the database, according to the
+    /// watermarks table. Returns `None` if any of the expected pipelines are missing data.
+    pub async fn latest_checkpoint(&self) -> anyhow::Result<Option<u64>> {
+        use watermarks::dsl as w;
+
+        let mut conn = self
+            .db
+            .connect()
+            .await
+            .context("Failed to connect to database")?;
+
+        let latest: HashMap<String, i64> = w::watermarks
+            .select((w::pipeline, w::checkpoint_hi_inclusive))
+            .filter(w::pipeline.eq_any(&self.pipelines))
+            .load(&mut conn)
+            .await?
+            .into_iter()
+            .collect();
+
+        for pipeline in &self.pipelines {
```

Comment:
would `latest.len() != pipelines.len()` suffice to return None

Reply:
Yes, I think so!